js_modules/dagster-ui/packages/ui-core/src/hooks/useThrottledMemo.tsx

```diff
@@ -0,0 +1,71 @@
+import {createContext, useContext, useEffect, useRef, useState} from 'react';
+import {unstable_batchedUpdates} from 'react-dom';
+
+/**
+ * Context to accumulate updates over 200ms and batch them together to reduce the number of renders
+ */
+export const ThrottledMemoBatchingContext = createContext<{enqueue: (update: () => void) => void}>(
+  (() => {
+    let isScheduled = false;
+    const queue: Array<() => void> = [];
+    return {
+      enqueue: (update: () => void) => {
+        queue.push(update);
+        if (!isScheduled) {
+          isScheduled = true;
+          setTimeout(() => {
+            unstable_batchedUpdates(() => {
+              while (queue.length) {
+                queue.shift()!();
+              }
+              isScheduled = false;
+            });
+          }, 200);
+        }
+      },
+    };
+  })(),
+);
+
+export const useThrottledMemo = <T,>(
+  factory: () => T,
+  deps: React.DependencyList,
+  delay: number,
+): T => {
+  const [state, setState] = useState<T>(factory);
+  const lastRun = useRef<number>(Date.now());
+  const timeoutId = useRef<NodeJS.Timeout | null>(null);
+  const {enqueue} = useContext(ThrottledMemoBatchingContext);
+
+  useEffect(() => {
+    const now = Date.now();
+    if (now - lastRun.current >= delay) {
+      enqueue(() => {
+        setState(factory);
+      });
+      lastRun.current = now;
+    } else {
+      if (timeoutId.current) {
+        clearTimeout(timeoutId.current);
+      }
+      timeoutId.current = setTimeout(
+        () => {
+          enqueue(() => {
+            setState(factory);
+          });
```

Comment:
Hmm I think this introduces an edge case where we call setState after the hook is unmounted, since `enqueue` can be equivalent to `setTimeout(() => setState(factory), 300)`

Reply:
ah good catch I'll add a guard for this